[NativeAOT] Assertions while building libraries tests

A few tests fail to compile in Checked build of libraries tests.
(examp[le log: https://[dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/344603/logs/705](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/344603/logs/705))



```
 2023-07-19T05:57:52.0500275Z   Process terminated. Assertion failed.
2023-07-19T05:57:52.0501327Z      at ILCompiler.Dataflow.MethodBodyScanner.InterproceduralScan(MethodIL startingMethodBody) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/MethodBodyScanner.cs:line 330
2023-07-19T05:57:52.0502397Z      at ILCompiler.Dataflow.ReflectionMethodBodyScanner.InterproceduralScan(MethodIL methodBody) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:line 103
2023-07-19T05:57:52.0503563Z      at ILCompiler.Dataflow.ReflectionMethodBodyScanner.ScanAndProcessReturnValue(NodeFactory factory, FlowAnnotations annotations, Logger logger, MethodIL methodBody) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:line 128
2023-07-19T05:57:52.0553815Z      at ILCompiler.DependencyAnalysis.DataflowAnalyzedMethodNode.GetStaticDependencies(NodeFactory factory) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/DataflowAnalyzedMethodNode.cs:line 32
2023-07-19T05:57:52.0557054Z      at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1 node) in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 182
2023-07-19T05:57:52.0557674Z      at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependencies(DependencyNodeCore`1 node) in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 222
2023-07-19T05:57:52.0558201Z      at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack() in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 257
2023-07-19T05:57:52.0558679Z      at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 308
2023-07-19T05:57:52.0559130Z      at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan() in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:line 140
2023-07-19T05:57:52.0559526Z      at ILCompiler.Program.<Run>g__RunScanner|4_0(<>c__DisplayClass4_0&) in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 467
2023-07-19T05:57:52.0559854Z      at ILCompiler.Program.Run() in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 447
2023-07-19T05:57:52.0560231Z      at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass221_0.<.ctor>b__0(ParseResult result) in /_/src/coreclr/tools/aot/ILCompiler/ILCompilerRootCommand.cs:line 277
2023-07-19T05:57:52.0560609Z      at System.CommandLine.Invocation.InvocationPipeline.Invoke(ParseResult parseResult)
2023-07-19T05:57:52.0560914Z      at ILCompiler.Program.Main(String[] args) in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 727
2023-07-19T05:57:52.1918929Z /__w/1/s/artifacts/bin/coreclr/linux.x64.Checked/build/Microsoft.NETCore.Native.targets(300,5): error MSB3073: The command ""/__w/1/s/artifacts/bin/coreclr/linux.x64.Checked/x64/ilc/ilc" @"/__w/1/s/artifacts/obj/Microsoft.Extensions.Logging.Tests/Release/net8.0/native/Microsoft.Extensions.Logging.Tests.ilc.rsp"" exited with code 134. [/__w/1/s/src/libraries/Microsoft.Extensions.Logging/tests/Common/Microsoft.Extensions.Logging.Tests.csproj::TargetFramework=net8.0]

```

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedNames.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedNames.cs
@@ -7,7 +7,7 @@ namespace ILCompiler.Dataflow
 {
     internal sealed class CompilerGeneratedNames
     {
-        internal static bool IsGeneratedMemberName(string memberName)
+        private static bool IsGeneratedMemberName(string memberName)
         {
             return memberName.Length > 0 && memberName[0] == '<';
         }
@@ -49,7 +49,7 @@ namespace ILCompiler.Dataflow
             return fieldName.Length > i + 1 && fieldName[i + 1] == '2';
         }
 
-        internal static bool IsGeneratedType(string name) => IsStateMachineType(name) || IsLambdaDisplayClass(name);
+        internal static bool IsStateMachineOrDisplayClass(string name) => IsStateMachineType(name) || IsLambdaDisplayClass(name);
 
         internal static bool IsLambdaOrLocalFunction(string methodName) => IsLambdaMethod(methodName) || IsLocalFunction(methodName);
 

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs
@@ -73,7 +73,7 @@ namespace ILCompiler.Dataflow
             internal TypeCache(MetadataType type, Logger? logger, ILProvider ilProvider)
             {
                 Debug.Assert(type == type.GetTypeDefinition());
-                Debug.Assert(!CompilerGeneratedNames.IsGeneratedMemberName(type.Name));
+                Debug.Assert(!CompilerGeneratedNames.IsStateMachineOrDisplayClass(type.Name));
 
                 Type = type;
 
@@ -194,7 +194,7 @@ namespace ILCompiler.Dataflow
                     if (TryGetStateMachineType(method, out MetadataType? stateMachineType))
                     {
                         Debug.Assert(stateMachineType.ContainingType == type ||
-                            (CompilerGeneratedNames.IsGeneratedMemberName(stateMachineType.ContainingType.Name) &&
+                            (CompilerGeneratedNames.IsStateMachineOrDisplayClass(stateMachineType.ContainingType.Name) &&
                              stateMachineType.ContainingType.ContainingType == type));
                         Debug.Assert(stateMachineType == stateMachineType.GetTypeDefinition());
 
@@ -321,7 +321,7 @@ namespace ILCompiler.Dataflow
                     MetadataType generatedType,
                     Dictionary<MetadataType, TypeArgumentInfo> generatedTypeToTypeArgs)
                 {
-                    Debug.Assert(CompilerGeneratedNames.IsGeneratedType(generatedType.Name));
+                    Debug.Assert(CompilerGeneratedNames.IsStateMachineOrDisplayClass(generatedType.Name));
                     Debug.Assert(generatedType == generatedType.GetTypeDefinition());
 
                     var typeInfo = generatedTypeToTypeArgs[generatedType];
@@ -361,7 +361,7 @@ namespace ILCompiler.Dataflow
                             else
                             {
                                 // Must be a type ref
-                                if (method.OwningType is not MetadataType owningType || !CompilerGeneratedNames.IsGeneratedType(owningType.Name))
+                                if (method.OwningType is not MetadataType owningType || !CompilerGeneratedNames.IsStateMachineOrDisplayClass(owningType.Name))
                                 {
                                     userAttrs = param;
                                 }
@@ -501,7 +501,7 @@ namespace ILCompiler.Dataflow
         {
             foreach (var nestedType in type.GetNestedTypes())
             {
-                if (!CompilerGeneratedNames.IsGeneratedMemberName(nestedType.Name))
+                if (!CompilerGeneratedNames.IsStateMachineOrDisplayClass(nestedType.Name))
                     continue;
 
                 yield return nestedType;
@@ -565,7 +565,7 @@ namespace ILCompiler.Dataflow
             // State machines can be emitted into display classes, so we may also need to go one more level up.
             // To avoid depending on implementation details, we go up until we see a non-compiler-generated type.
             // This is the counterpart to GetCompilerGeneratedNestedTypes.
-            while (userType != null && CompilerGeneratedNames.IsGeneratedMemberName(userType.Name))
+            while (userType != null && CompilerGeneratedNames.IsStateMachineOrDisplayClass(userType.Name))
                 userType = userType.ContainingType as MetadataType;
 
             if (userType is null)
@@ -607,7 +607,7 @@ namespace ILCompiler.Dataflow
         public IReadOnlyList<GenericParameterDesc?>? GetGeneratedTypeAttributes(MetadataType type)
         {
             MetadataType generatedType = (MetadataType)type.GetTypeDefinition();
-            Debug.Assert(CompilerGeneratedNames.IsGeneratedType(generatedType.Name));
+            Debug.Assert(CompilerGeneratedNames.IsStateMachineOrDisplayClass(generatedType.Name));
 
             var typeCache = GetCompilerGeneratedStateForType(generatedType);
             if (typeCache is null)

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/FlowAnnotations.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/FlowAnnotations.cs
@@ -641,7 +641,7 @@ namespace ILLink.Shared.TrimAnalysis
 
             private IReadOnlyList<GenericParameterDesc?>? GetGeneratedTypeAttributes(EcmaType typeDef)
             {
-                if (!CompilerGeneratedNames.IsGeneratedType(typeDef.Name))
+                if (!CompilerGeneratedNames.IsStateMachineOrDisplayClass(typeDef.Name))
                 {
                     return null;
                 }

--- a/src/tools/illink/src/linker/Linker.Dataflow/CompilerGeneratedNames.cs
+++ b/src/tools/illink/src/linker/Linker.Dataflow/CompilerGeneratedNames.cs
@@ -5,7 +5,7 @@ namespace Mono.Linker.Dataflow
 {
 	sealed class CompilerGeneratedNames
 	{
-		internal static bool IsGeneratedMemberName (string memberName)
+		private static bool IsGeneratedMemberName (string memberName)
 		{
 			return memberName.Length > 0 && memberName[0] == '<';
 		}
@@ -47,7 +47,7 @@ namespace Mono.Linker.Dataflow
 			return fieldName.Length > i + 1 && fieldName[i + 1] == '2';
 		}
 
-		internal static bool IsGeneratedType (string name) => IsStateMachineType (name) || IsLambdaDisplayClass (name);
+		internal static bool IsStateMachineOrDisplayClass (string name) => IsStateMachineType (name) || IsLambdaDisplayClass (name);
 
 		internal static bool IsLambdaOrLocalFunction (string methodName) => IsLambdaMethod (methodName) || IsLocalFunction (methodName);
 

--- a/src/tools/illink/src/linker/Linker.Dataflow/CompilerGeneratedState.cs
+++ b/src/tools/illink/src/linker/Linker.Dataflow/CompilerGeneratedState.cs
@@ -43,7 +43,7 @@ namespace Mono.Linker.Dataflow
 		static IEnumerable<TypeDefinition> GetCompilerGeneratedNestedTypes (TypeDefinition type)
 		{
 			foreach (var nestedType in type.NestedTypes) {
-				if (!CompilerGeneratedNames.IsGeneratedMemberName (nestedType.Name))
+				if (!CompilerGeneratedNames.IsStateMachineOrDisplayClass (nestedType.Name))
 					continue;
 
 				yield return nestedType;
@@ -114,7 +114,7 @@ namespace Mono.Linker.Dataflow
 			// State machines can be emitted into display classes, so we may also need to go one more level up.
 			// To avoid depending on implementation details, we go up until we see a non-compiler-generated type.
 			// This is the counterpart to GetCompilerGeneratedNestedTypes.
-			while (type != null && CompilerGeneratedNames.IsGeneratedMemberName (type.Name))
+			while (type != null && CompilerGeneratedNames.IsStateMachineOrDisplayClass (type.Name))
 				type = type.DeclaringType;
 
 			if (type is null)
@@ -206,7 +206,7 @@ namespace Mono.Linker.Dataflow
 
 				if (TryGetStateMachineType (method, out TypeDefinition? stateMachineType)) {
 					Debug.Assert (stateMachineType.DeclaringType == type ||
-						CompilerGeneratedNames.IsGeneratedMemberName (stateMachineType.DeclaringType.Name) &&
+						CompilerGeneratedNames.IsStateMachineOrDisplayClass (stateMachineType.DeclaringType.Name) &&
 						 stateMachineType.DeclaringType.DeclaringType == type);
 					callGraph.TrackCall (method, stateMachineType);
 
@@ -324,7 +324,7 @@ namespace Mono.Linker.Dataflow
 				Dictionary<TypeDefinition, TypeArgumentInfo> generatedTypeToTypeArgs,
 				LinkContext context)
 			{
-				Debug.Assert (CompilerGeneratedNames.IsGeneratedType (generatedType.Name));
+				Debug.Assert (CompilerGeneratedNames.IsStateMachineOrDisplayClass (generatedType.Name));
 
 				var typeInfo = generatedTypeToTypeArgs[generatedType];
 				if (typeInfo.OriginalAttributes is not null) {
@@ -353,7 +353,7 @@ namespace Mono.Linker.Dataflow
 							} else {
 								// Must be a type ref
 								var owningRef = (TypeReference) owner;
-								if (!CompilerGeneratedNames.IsGeneratedType (owningRef.Name)) {
+								if (!CompilerGeneratedNames.IsStateMachineOrDisplayClass (owningRef.Name)) {
 									userAttrs = param;
 								} else if (context.TryResolve ((TypeReference) param.Owner) is { } owningType) {
 									MapGeneratedTypeTypeParameters (owningType, generatedTypeToTypeArgs, context);
@@ -443,7 +443,7 @@ namespace Mono.Linker.Dataflow
 		/// </summary>
 		public IReadOnlyList<ICustomAttributeProvider>? GetGeneratedTypeAttributes (TypeDefinition generatedType)
 		{
-			Debug.Assert (CompilerGeneratedNames.IsGeneratedType (generatedType.Name));
+			Debug.Assert (CompilerGeneratedNames.IsStateMachineOrDisplayClass (generatedType.Name));
 
 			var typeToCache = GetCompilerGeneratedStateForType (generatedType);
 			if (typeToCache is null)

--- a/src/tools/illink/src/linker/Linker.Dataflow/FlowAnnotations.cs
+++ b/src/tools/illink/src/linker/Linker.Dataflow/FlowAnnotations.cs
@@ -399,7 +399,7 @@ namespace ILLink.Shared.TrimAnalysis
 
 		private IReadOnlyList<ICustomAttributeProvider>? GetGeneratedTypeAttributes (TypeDefinition typeDef)
 		{
-			if (!CompilerGeneratedNames.IsGeneratedType (typeDef.Name)) {
+			if (!CompilerGeneratedNames.IsStateMachineOrDisplayClass (typeDef.Name)) {
 				return null;
 			}
 			var attrs = _context.CompilerGeneratedState.GetGeneratedTypeAttributes (typeDef);

